File: DesignPatterns/01_GOF23/DP_examples/CSharp/03_Behavioral/State/State.cs
namespace Behavioral.State;
    /// <summary>
    /// TCP 状态的缺省行为
    /// </summary>
    internal abstract class TCPState {
        public virtual void Connect(TCPConnection connection) { }
        public virtual void Receive(TCPConnection connection, out byte[]? buffer) => buffer = null;
        public virtual void Close(TCPConnection connection) { }
        public virtual void SendMessage(TCPConnection connection, byte[] buffer) { }
        protected virtual void ChangeState(TCPConnection connection, TCPState state) {
            Console.WriteLine($"Switch state to >> {state}");
            connection.ChangeState(state);
        }
}

File: 01_GOF23/DP_examples/C++/03_Behavioral/State/State.cpp
#include "State.hpp"
#include "Context.hpp"

// 实现ChangeState方法
void TCPState::ChangeState(TCPConnection* connection, TCPState* state) {
    std::cout << "Switch state to >> " << typeid(*state).name() << std::endl;
    connection->ChangeState(state);
}
